deepchem/utils/poly_wd_graph_utils.py

```diff
@@ -0,0 +1,153 @@
+        max_wildcard = max(
+            [int(x) for x in self._get_all_wildcards(monomer_mols)])
+        for wildcard in range(1, max_wildcard + 1):
+            if str(wildcard) + "*" not in monomer_mols:
+                raise ValueError(
+                    f"Invalid wildcard format: The wildcard {wildcard} is not present in the monomer molecules string  as per the sequence of the maximum {max_wildcard}!"
+                )
+
+    def _validate_polymer_rules(self, datapoint: str):
+        monomer_mols, _, polymer_rules = self.get_parsed_vals(datapoint)
+        polymer_rule_list = self.get_polymer_rules(polymer_rules)
+        for rules in polymer_rule_list:
+            splits = rules.split(":")
+            if len(splits) != 3:
+                raise ValueError(
+                    f"Invalid polymer rule format: The rule must contain exactly 3 splits ! but found {len(splits)} splits"
+                )
+            if "-" not in splits[0]:
+                raise ValueError(
+                    f"Invalid polymer rule format: The bond string between two wildcard index must be seprated by '-', got invalid data {splits[0]}"
+                )
+            elif len(splits[0].split("-")) != 2 and any(
+                    elem == "" for elem in splits[0].split("-")):
+                raise ValueError(
+                    f"Invalid polymer rule format: The first split must contain exactly 2 splits to depict connection between atom indexes! but found {len(splits[0].split('-'))} splits"
+                )
+            else:
+                for wild_card_index in splits[0].split("-"):
+                    if not wild_card_index.isdigit():
+                        raise ValueError(
+                            f"Invalid polymer rule format: The first split must contain only digits! but found {wild_card_index}"
+                        )
+                    if wild_card_index not in monomer_mols:
+                        raise ValueError(
+                            f"Invalid polymer rule format: The first split must contain only valid wild card indexes! but found {wild_card_index} which is not in {monomer_mols}"
+                        )
+
+    def validate(self, datapoint: str):
+        """
+        This method validates the string format of weighted directed graph data.
+        To validate the string format it checks for following conditions:
+
+        1. The number of fragments and the number of monomer molecules should match.
+        2. The wild card indexes should be present in the monomer molecules string and should be in the correct sequence.
+        3. The polymer rules should be in the correct format.
+        4. The atom indexes in the polymer rules should be valid and present in the monomer molecules string.
+
+        It raises ValueError if the string format is invalid.
+
+        Parameters
+        ----------
+        datapoint : str
+            The datapoint string to validate
+        Returns
```

Comment:
pls leave an empty line between sections. it helps the deepchem docs render properly.